streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```
@@ -606,14 +609,19 @@ private boolean removeFailedTask(final TaskId taskId, final CompletableFuture<Re
         private void pauseTask(final Task task) {
             final TaskId taskId = task.id();
             // do not need to unregister changelog partitions for paused tasks
-            measureCheckpointLatency(() -> task.maybeCheckpoint(true));
-            pausedTasks.put(taskId, task);
-            updatingTasks.remove(taskId);
-            if (task.isActive()) {
-                transitToUpdateStandbysIfOnlyStandbysLeft();
+            try {
+                measureCheckpointLatency(() -> task.maybeCheckpoint(true));
+                pausedTasks.put(taskId, task);
+                updatingTasks.remove(taskId);
+                if (task.isActive()) {
+                    transitToUpdateStandbysIfOnlyStandbysLeft();
+                }
+                log.info((task.isActive() ? "Active" : "Standby")
+                    + " task " + task.id() + " was paused from the updating tasks and added to the paused tasks.");
+
+            } catch (final StreamsException streamsException) {
```

It seems we are adding a similar `try-catch` on multiple places. Should we instead add a new catch block to `run()` method to have it in a single place (and also avoid that this PR might miss some place where it should get added otherwise, too)?

---

the idea was to handle one task at a time instead of failing all of them together. so if we have 3 tasks, one fails but the other 2 succeed, we will have only one failed task instead of all 3.
similarly to https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java#L363 and https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java#L516

---

I see. Hard to read on the PR, as there is no loop here...

But I am not even sure why we would want to fail a task if we cannot checkpoint for this case? We only `pause` a task, and it seems to be non-fatal if we fail to checkpoint? Wondering if we should swallow the exception entirely and just keep the task w/o the need to move it into error state and hand back to `StreamsThreads`?

\cc @lucasbru

---

Hmm, the comment of maybeCheckpoint mentions this:

```
     * @throws TaskMigratedException recoverable error sending changelog records that would cause the task to be removed
     * @throws StreamsException fatal error when flushing the state store, for example sending changelog records failed
     *                          or flushing state store get IO errors; such error should cause the thread to die
```

That would actually indicate that we have to treat streams exception fatally here. But I haven't yet tracked down where they come from and if they can happen in the state updater. Maybe that comment should be made more precise? But going by the commen, we actually need to fail here.

---

>But I haven't yet tracked down where they come from and if they can happen in the state updater

the problem I am trying to address by this PR is pretty much that `maybeCheckpoint` threw an exception when was called by the state updater. Which caused the state updater thread to die and the stream thread shutdown to hang.
even though it happened during `maybeCheckpointTasks` not during `pauseTasks`, but it may as well.

`maybeCheckpoint` threw an exception because the RockDB failed to flush(I assume due to some issues with the file system)

---

The question is
 - Where does the `TaskMigratedException` come from. Can it happen, or should be update the javadoc?
 - Why does the comment indicate that the thread should die on an I/O error? @mjsax any idea? If the comment is correct, we need to call handleStreamsExceptionWithoutTask to handle the exception. If not, we should change the comment.

---

>Where does the TaskMigratedException come from. Can it happen, or should be update the javadoc?

I wasn't able to find a way to receive `TaskMigratedException` from`maybeCheckpoint`, as `maybeCheckpoint` doesn't really try to produce any messages. however, as it's a child of `StreamsException` it's possible that a custom implementation of a store may throw one(although I can't think of a reason).

> Why does the comment indicate that the thread should die on an I/O error?

it's the default behavior: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java#L453, it's not unique for `maybeCheckpoint`

---

Thanks for reading the JavaDoc comments Lucas. This makes it pretty clear. `sending changelog records failed` is fatal as it would be data loss, and we cannot proceed. It make sense after the fact -- before we can write the local checkpoint file, we need to flush the producer (missed this point originally).

> Why does the comment indicate that the thread should die on an I/O error

For such critical errors, we don't want to have any built-in recovery mechanism (design decision), but let the user opt into auto-recovery via the uncaught exception handler.

As we flush pending producer writes, a `TaskMigratedException` can originate from this flush. We should still handle this one gracefully, and not let the thread die.

---

Wait, you are saying a `StreamsException` which is not `TaskMigrationException` should cause a fatal error, no? Because this is not what the PR is doing now.

---

I think the comment by the thread means the streams thread, not the state updater thread.
because the code I shared above does pretty much that - if we get a StreamsException, we shut down the streams thread

---

Oh! Sorry. My fault. You are right, I misremembered what we are doing with the expections downstream. All good!